refactor: make `hex_as_bytes` DRY

dhilipsiva opened the pull request:
- Modify `hex_as_bytes` function to return an array of bytes based on input length
- Remove mutable `output` parameter in `hex_as_bytes` function
- Update usage of `hex_as_bytes` function to use generic type and constant size

dhilipsiva committed:
refactor: make `hex_as_bytes` DRY

- Modify `hex_as_bytes` function to return an array of bytes based on input length
- Remove mutable `output` parameter in `hex_as_bytes` function
- Update usage of `hex_as_bytes` function to use generic type and constant size
- Simplify the conversion of `DtlsFingerprint` to `web_rtc_transport::Fingerprint`
- Simplify the conversion of `web_rtc_transport::Fingerprint` to `DtlsFingerprint`

ibc commented:
Please run `cargo fmt`, `cargo clippy` and `cargo test` and commit changes. CI is failing.

nazar-pc commented:
Just a note: this will bump required Rust compiler version to one that supports const generics (which is relatively old at this point, but still)

ds-colligence commented:
@ibc I ran all of those 3 commands before pushing the code. `cargo fmt` did not report any bugs. All tests are passing. cliipy does report a few things, but it is not from the code that I touched, it was already there. Should I be fixing those as well?

dhilipsiva commented:
@ibc I ran all of those 3 commands before pushing the code. cargo fmt did not report any changes. All tests are passing. clippy does report a few things, but it is not from the code that I touched, it was already there. Should I be fixing those as well?

ibc commented:
Looks like CI or you are running different Rust versions. 

@nazar-pc, how to proceed here? AFAIU mediasoup-rust mandates a minimal Rust version but the user is free to use a major one. Does mediasoup-rust fail to build is running Rust version doesn't satisfy that minimum version declared somewhere in mediasoup files? Should we update that minimum version and match whatever version runs in CI servers?

nazar-pc commented:
We have `rust-toolchain.toml` in the repo. Assuming Rust toolchain is installed with rustup and Cargo is used, it should be used by default unless overridden. Don't override it, please, use what it chooses out of the box.

P.S. I don't mind bumping version in `rust-toolchain.toml` to much newer release.

ibc commented:
> We have `rust-toolchain.toml` in the repo. Assuming Rust toolchain is installed with rustup and Cargo is used, it should be used by default unless overridden.

I don't understand this. `rustup` will update Rust to whatever the newest version is, not to the version we say in `rust-toolchain.toml`.

ibc commented:
@dhilipsiva can you run `rustup --version` in your system? It prints this in my computer:

```
$ rustup --version

rustup 1.26.0 (5af9b9484 2023-04-05)
info: This is the version for the rustup toolchain manager, not the rustc compiler.
info: The currently active `rustc` version is `rustc 1.72.0 (5680fa18f 2023-08-23)`
```

Which matches the Rust version declared in `rust-toolchain.toml` in mediasoup (1.72.0). And when I run all `cargo` commands (fmt, clippy, etc) I do not see any warning, so probably you are using another version of Rust.

dhilipsiva commented:
@nazar-pc @ibc 

I just realized I have been using rust toolchain installed using brew.

```
rustc --version
rustc 1.78.0 (9b00956e5 2024-04-29) (Homebrew)
```

nazar-pc commented:
> I don't understand this. `rustup` will update Rust to whatever the newest version is, not to the version we say in `rust-toolchain.toml`.

If `rust-toolchain.toml` is present, that specified toolchain version will be used for the project. Even if default on the system is newer or older. That is the whole point of that file: to have predictable toolchain and set of its components. If you run `rustc --version` in project directory you'll get the same version as in `rust-toolchain.toml`.

dhilipsiva commented:
@ibc 

I just installed my toolchain using `rustup`.

```
rustup --version
rustup 1.27.1 (54dd3d00f 2024-04-24)
info: This is the version for the rustup toolchain manager, not the rustc compiler.
info: The currently active `rustc` version is `rustc 1.79.0 (129f3b996 2024-06-10)`

rustc --version
rustc 1.79.0 (129f3b996 2024-06-10)
```

* `cargo fmt` reported/changed nothing
* `cargo test` passed
* `cargo clippy` reported parts of code I never touched.

dhilipsiva commented:
@ibc @nazar-pc I ran all the commands from the root of the repo. Should I have run them from `rust` folder instead?

nazar-pc commented:
> @ibc @nazar-pc I ran all the commands from the root of the repo. Should I have run them from `rust` folder instead?

If you do that in project directory you should get this:
```
nazar-pc@nazar-pc:/w/g/mediasoup> rustc --version
rustc 1.72.0 (5680fa18f 2023-08-23)
```

Even though global version is different:
```
nazar-pc@nazar-pc:~> rustc --version
rustc 1.77.2 (25ef9e3d8 2024-04-09)
```

Make sure your brew version is not overriding rustup's version.

nazar-pc commented:
Root works just fine as well

dhilipsiva commented:
Can confirm that I have uninstalled brew version of rust before running the cargo commands.

```
which rustc
/home/dhilipsiva/.cargo/bin/rustc
which rustup
/home/dhilipsiva/.cargo/bin/rustup
```

nazar-pc commented:
well, just bump `rust-toolchain.toml` to 1.79.0 then, it'll match your current verison and everything should be fine.

ibc commented:
> If `rust-toolchain.toml` is present, that specified toolchain version will be used for the project. Even if default on the system is newer or older. That is the whole point of that file: to have predictable toolchain and set of its components. If you run `rustc --version` in project directory you'll get the same version as in `rust-toolchain.toml`.

Imagine I have rustc 1.72.0 in my system and `rust-toolchain.toml` says 1.79.0. Obviously things are gonna fail due to version mismtach.

Imagine I have rustc 1.79.0 in my system and `rust-toolchain.toml` says 1.72.0. What would happen? I don't have rustc 1.72.0. installed in my system.

nazar-pc commented:
> > If `rust-toolchain.toml` is present, that specified toolchain version will be used for the project. Even if default on the system is newer or older. That is the whole point of that file: to have predictable toolchain and set of its components. If you run `rustc --version` in project directory you'll get the same version as in `rust-toolchain.toml`.
> 
> Imagine I have rustc 1.72.0 in my system and `rust-toolchain.toml` says 1.79.0. Obviously things are gonna fail due to version mismtach.
> 
> Imagine I have rustc 1.79.0 in my system and `rust-toolchain.toml` says 1.72.0. What would happen? I don't have rustc 1.72.0. installed in my system.

It will not fail in either case, It'll install (if not installed already) correct version and will use correct version in the project and nothing else.

dhilipsiva commented:
> well, just bump `rust-toolchain.toml` to 1.79.0 then, it'll match your current verison and everything should be fine.

Sorry. Should I update the version, commit and push to this branch? Is that what you meant? :sweat_smile:

nazar-pc commented:
Yes

dhilipsiva committed:
chore: Update Rust toolchain channel to version 1.79.0

- Update Rust toolchain channel from "1.72.0" to "1.79.0" in the [rust-toolchain.toml] file.

dhilipsiva committed:
chore: Update Rust toolchain channel to version 1.79.0

- Update Rust toolchain channel from "1.72.0" to "1.79.0" in the [rust-toolchain.toml] file.

dhilipsiva commented:
Update rust version in rust-toolchain.toml to 1.79.0

dhilipsiva committed:
Merge branch 'v3' into dry-DtlsFingerprint-from_fbs

* v3:
  chore: Update Rust toolchain channel to version 1.79.0

dhilipsiva committed:
chore: apply linting

ibc commented:
Let's wait for CI.

ibc commented:
CI failing in `cargo clippy`: https://github.com/versatica/mediasoup/actions/runs/9610107155/job/26507750935?pr=1409

ibc commented:
> CI failing in `cargo clippy`: https://github.com/versatica/mediasoup/actions/runs/9610107155/job/26507750935?pr=1409

I assume those are due to new version of Rust.

dhilipsiva committed:
Merge branch 'v3' into dry-DtlsFingerprint-from_fbs

* v3:
  chore: Update Rust toolchain channel to version 1.79.0
  Fix issue #1374
  cosmetic
  fix Simulcast IncreaseLayer bug when producer score is zero (#1410)
  Update NPM deps
  TcpConnectionHandle.cpp: properly close handle
  fix asan error for new-delete-type-mismatch (#1411)

dhilipsiva committed:
chore: clippy autofix

cargo clippy --fix --lib -p mediasoup

dhilipsiva committed:
chore: cargo fmt

dhilipsiva committed:
chore: fix clippy warnings

- Added `#[allow(dead_code)]` attribute to `ChannelReadCallback` in `channel_read_fn.rs`
- Modified and added test cases in `webrtc_server.rs`
- Made changes to the `Producer` struct and related functions in `producer.rs`
- Made updates to the `ScalabilityMode` implementation in `scalability_modes.rs`
- Implemented various functions and callbacks in `router.rs`
- Made modifications and fixes in `ortc/tests.rs`
- Updated test cases in `webrtc_transport.rs`
- Added attributes to struct definitions in `channel_write_fn.rs` and `consumer.rs`
- Modified the code in `worker/build.rs`

dhilipsiva commented:
@ibc @nazar-pc Most of the clippy errors were due to planus codegen. I have updated the build script to include `#![allow(clippy::all)` within the generated module to exclude it from clippy. I have done so  using a simple string replace method in the worker build.rs script. I hope this is an acceptable solution. 

Edit: can confirm that all clippy warnings are resolved when tested on my local.

dhilipsiva commented on worker/build.rs:
```diff
@@ -33,12 +33,13 @@ fn main() {
+        .expect("Failed to generate Rust code from flatbuffers");
+
+    let planus_content =
+        planus_content.replace("mod root {", "mod root {\n    #![allow(clippy::all)]");
```
@ibc @nazar-pc I hope this is an acceptable solution.

nazar-pc commented:
Well, the goal is not to get clippy to shut up by suppressing lints/commenting code without any explanation, but to actually address it.

I wanted to fix it, but you have sent PR from an org rather than personal account, so I can't do that. I have pushed commit with fixes here: https://github.com/nazar-pc/mediasoup/tree/dry-DtlsFingerprint-from_fbs-fix-lints

dhilipsiva commented:
Oh, wow! Using Drop to silence the dead_code error is such a neat trick. :) It didn't cross my mind.

nazar-pc commented:
Well, another way is to actually use named field (rather than tuple), where name starts with `_`, that way clippy will not complain either. Probably should have done that instead.

nazar-pc commented:
The key though is to leave an explanation why that is okay/expected

ibc commented:
So is this ready to merge?

nazar-pc commented:
I expect one more update as described above

dhilipsiva commented:
Oh sincere apologies. I misunderstood. I mistakenly assumed that you will merge the changes you made on your fork. Let me do this in a couple of hours.

dhilipsiva committed:
refactor: Refactor channel read and write callbacks

- Add a new method `new` to the `ChannelReadCallback` struct in `channel_read_fn.rs`
- Modify the implementation of the `ChannelReadCallback` struct to use the new method
- Update the `prepare_channel_read_fn` function to use the `new` method of `ChannelReadCallback`
- Add a typedef `CallbackType` to `channel_write_fn.rs` to avoid warnings
- Refactor the `ChannelReadCallback` struct in `channel_write_fn.rs` to use `CallbackType`
- Add a constructor method `new` to `ChannelReadCallback` in `channel_write_fn.rs`
- Update the `prepare_channel_write_fn` function to use the new `ChannelReadCallback` constructor in `channel_write_fn.rs`
- Modify the `RtpCodecCapability` struct in `consumer.rs` to include additional parameters for audio and video codecs
- Add `RtpHeaderExtension` structs for various media kinds and URIs in `consumer.rs`
- Refactor the `ExecutorGuard` struct in `consumer.rs` to use a new `new` method instead of the `impl` block
- Modify the `create_executor` function in `consumer.rs` to use the new `ExecutorGuard` initialization
- Modify the `init` function in `consumer.rs` to return multiple values rather than a single tuple
- Remove the `#[allow(dead_code)]` attribute from the `impl RtpParameters` block in `rtp_parameters.rs`

dhilipsiva commented:
I have removed ignoring dead_code warnings. I have used named fields as suggested. Please let me know if there is anything I need to change :)

dhilipsiva commented:
While going through the commit @nazar-pc made on their personal repo, I realized I missed something: 

Silencing clippy lints on files generated by plaun codegen: https://github.com/nazar-pc/mediasoup/commit/7cdf64865afb63f61c68490bf0a20a887e937a42#diff-4bce653f545fe2edb899732221932e371b14c992e8304ab8f6a9677a142a7ec8R4

I did this same exact change and I can confirm that this is what caused #1413 

```
../../../../../../worker/meson.build:284:20: ERROR: Subproject exists but has no meson.build file.
```

I don't think I am smart enough to figure this one out. I deleted the registry folder as mentioned in #1413' s thread - but the same error kept happening.

nazar-pc commented:
During development files are stored in the project's directory. Delete everything in `worker/subprojects` except wrap files and you should be good to go. `make clean-all` should work in `worker` directory too.

ibc commented:
Now this is ready to merge, right?

nazar-pc commented on rust/src/worker/utils/channel_read_fn.rs:
```diff
@@ -8,9 +8,17 @@ unsafe extern "C" fn free_vec(message: *mut u8, message_len: u32, message_capaci
-    Box<dyn (FnMut(UvAsyncT) -> Option<Vec<u8>>) + Send + 'static>,
-);
+pub(super) struct ChannelReadCallback {
+    _callback: Box<dyn (FnMut(UvAsyncT) -> Option<Vec<u8>>) + Send + 'static>,
```
I'd really like for comments to be present at such locations

nazar-pc reviewed (commented):
Not all comments/suggestions were ported into here still, Inaki

dhilipsiva committed:
refactor: fix planus codegen clippy silencing

dhilipsiva committed:
chore: update docs

ds-colligence commented:
I have made all required changes. Please let me know if I have missed any. If there are no more changes and the tests succeeds, we can merge this.

nazar-pc reviewed (approved):
Looks good, thank you for contribution and patience!

ibc commented:
Merging, thanks a lot.

ibc commented:
I've updated Rust CHANGELOG for next release: https://github.com/versatica/mediasoup/commit/a80630aa107204fee677e4cf3b11f0f7b22fb064
